doc: kernel: fix k_poll example

clamattia opened the pull request:
Fix `k_poll()` loop example by using separate `if` blocks instead of `if`-`else`-chain to ensure all events are handled, even if multiple are active at the same time.

Signed-off-by: Cla Mattia Galliard
<cla-mattia.galliard@zuehlke.com>

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/76685

peter-mitsis commented:
The proposed changes look good--it is just the failing compliance checks that is holding back a +1 from me.

clamattia commented:
> The proposed changes look good--it is just the failing compliance checks that is holding back a +1 from me.

Thank you. I tried to address those.

andyross commented on doc/kernel/services/polling.rst:
```diff
@@ -185,15 +185,20 @@ to :c:macro:`K_POLL_STATE_NOT_READY` by the user.
         for(;;) {
             rc = k_poll(events, ARRAY_SIZE(events), K_FOREVER);
+            // NOTE: do not use an if-else-chain, because multiple events
+            // may be active simultaneously
```
Suggest removing this comment, or reword to be less prescriptive.  The existing code isn't actually incorrect, at worst it's inefficient as it requires a separate trip through k_poll() for each handled event (but of course in practice truly simultaneous wakeups are likely to be rare, or else reflect existing app inefficiencies with the "event model" or whatever).

clamattia replied:
Ah, I think, I start to understand the argument. Are you saying, that there will be a round-trip to `k_poll`, because the underlying object is still ready (semaphore given, data available in queue, signal pending, ...)? That makes sense.

Of course it requires, that the app only for example resets the poll-signal in the case it handled that event, which is kind of obvious.

Will remove the comment.

andyross reviewed (commented):
One nitpick but on the whole this looks correct to me.

peter-mitsis commented:
I think I know what is going on with the compliance check failure. Although the "Signed-off-by:" has been fixed, I think that the compliance check script is detecting an error in the "Author:" portion of the commit (which unfortunately is not shown in github, but can be seen using `git log HEAD^!`).

To fix that, try using `github commit --amend --reset-author` .

Commit:
doc: kernel: fix k_poll example

Fix `k_poll()` loop example by using separate `if` blocks instead of
`if`-`else`-chain to ensure all events are handled, even if multiple are
active at the same time.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

cfriedt commented:
One thing that I just noticed is that the code block being edited mostly duplicates the code block directly above, and also, that the same problem exists in the code block above.

Would it make sense to merge this code block with the one above, to reduce verbosity?

clamattia commented:
> One thing that I just noticed is that the code block being edited mostly duplicates the code block directly above, and also, that the same problem exists in the code block above.
> 
> Would it make sense to merge this code block with the one above, to reduce verbosity?

If so, I would remove the example above. Polling in a loop is probably the more common use-case. It is a critical bug to not reset the state then, which happened to me actually at first.

After comment from @andyross above , I understand my misunderstanding of the code and see, that the example code is actually correct. An alternative could maybe be to explain why the original code was correct in a bit more detail in the documentation. Or changing the `if else`- chain, like I propose is also fine.

andyross commented:
I think I caused a bikeshed, apologies.  To clarify the **code** here is better than what we have in the existing docs and should be merged.  I was only quibbling with the **comment** which implied that the earlier technique was "incorrect" when it was really just "inefficient".

clamattia commented:
> I think I caused a bikeshed, apologies. To clarify the **code** here is better than what we have in the existing docs and should be merged. I was only quibbling with the **comment** which implied that the earlier technique was "incorrect" when it was really just "inefficient".

Thanks for the clarification. I have removed the original **comment** after understanding.
